File: src/bin/0019-remove-nth-node-from-end-of-list.cpp
// @leetup=info id=19 lang=cpp slug=remove-nth-node-from-end-of-list

#include <algorithm>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// @leetup=code

class Solution {
 public:
  ListNode* removeNthFromEnd(ListNode* head, int n) {
    ListNode dummy_head(0, head);

    ListNode* a = &dummy_head;
    ListNode* b = &dummy_head;

    // Advance `a`.
    for (int i = 0; i < n; ++i) {
      a = a->next;
    }
    while (true) {
      if (!a->next) {
        b->next = b->next->next;
        return dummy_head.next;
      }
      a = a->next;
      b = b->next;
    }
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0019-remove-nth-node-from-end-of-list.py
# @leetup=info id=19 lang=python3 slug=remove-nth-node-from-end-of-list


from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @leetup=code
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy_head = ListNode(0, head)
        a = dummy_head
        b = dummy_head
        for _ in range(n):
            a = a.next
        while True:
            if not a.next:
                b.next = b.next.next
                return dummy_head.next
            a = a.next
            b = b.next


# @leetup=code
